### PR TITLE
Applab: Replace setInterval() with timedLoop()

### PR DESCRIPTION
**Pull request by islemaster**
Removes `setInterval()` and `clearInterval()` from the default App Lab toolbox, and replaces them with `timedLoop()` and `stopTimedLoop()`.

Back in https://github.com/code-dot-org/code-dot-org/pull/13104 we introduced `timedLoop` as a helper in the Maker category with a slightly different contract; now the curriculum team has decided they'd like to simplify it further and make it available in place of `setInterval` in App Lab (note that `setInterval` and `clearInterval` are still available, they're just not in the toolbox).

The new methods look like this:

image

> `intervalKey = timedLoop(ms, callback)`
>
> * **ms** - milliseconds between invocations of callback
> * **callback** - function called every (ms) milliseconds
> * Returns an interval key (number) just like `setInterval`.

> `stopTimedLoop([key])`
> * **key (optional)** - Which loop to stop. If omitted, stops _all_ intervals started by timedLoop().

That last bit about stopping _all_ intervals is the key reason to use this over `setInterval` - it lets beginners rely on a hidden global assuming (as the current lessons do) that we won't use more than one loop:

```javascript
timedLoop(1000, function () {
  console.log("tick");
});

button("button1", "Stop");
onEvent("button1", "click", function() {
  stopTimedLoop();
});
```

# Testing
Lots of changes to unit tests:

* Wrote a set of tests to pin the behavior of `timeoutList.js` before making changes to it, then added tests to cover new functionality added for `timedloop()` and `stopTimedLoop()`.
* Moved existing `timedLoop()` tests out of test/unit/lib/kits/maker and into test/unit/applab.  Had to introduce applab/dropletConfigTest.js to do this.

New unit tests:
```
  timeoutList
    setTimeout(fn, time)
      ✔ sets a timeout
      ✔ only happens once
    clearTimeout(id)
      ✔ stops a timeout from happening
      ✔ is a no-op if called after timeout occurs
      ✔ clears only one timeout according to given key
    clearTimeouts()
      ✔ stops all timeouts from happening
      ✔ is a no-op if no timeouts are set
    setInterval(fn, time)
      ✔ sets an interval
      ✔ happens every [interval] ms
    clearInterval(id)
      ✔ stops the interval from happening
      ✔ is a no-op if called on a bad key
      ✔ clears only one interval according to given key
    clearIntervals()
      ✔ stops all intervals
      ✔ also clears timedLoops
      ✔ is a no-op if no intervals are set
    timedLoop(interval, fn)
      ✔ runs code on an interval
      ✔ happens every [interval] ms
    stopTimedLoop([key])
      ✔ stops the loop from happening
      ✔ is a no-op if called when no timed loop is running
      ✔ is a no-op if called on a bad key
      ✔ clears only one loop if given a key
      ✔ clears all loops if called with no arguments
      ✔ does not clear setInterval calls if called with no arguments
```

Also manually tested with the following sample App Lab project:

```javascript
var blinking = false;
var lit = false;

button("button1", "Start");
setProperty("button1", "text-color", "white");
onEvent("button1", "click", function() {
  if (blinking) {
    blinking = false;
    stopTimedLoop();
    setText("button1", "Start");
  } else {
    blinking = true;
    timedLoop(200, function() {
      if (lit) {
        lit = false;
        setProperty("button1", "background-color", "black");
      } else {
        lit = true;
        setProperty("button1", "background-color", "green");
      }
    });
    setText("button1", "Stop All");
  }
});


// Second button
var blinking2 = false;
var lit2 = false;
var loopKey2;
button("button2", "Start");
setProperty("button2", "text-color", "white");
onEvent("button2", "click", function() {
  if (blinking2) {
    stopTimedLoop(loopKey2);
    setText("button2", "Start");
  } else {
    loopKey2 = timedLoop(200, function() {
      if (lit2) {
        setProperty("button2", "background-color", "black");
      } else {
        setProperty("button2", "background-color", "blue");
      }
      lit2 = !lit2;
    });
    setText("button2", "Stop One");
  }
  blinking2 = !blinking2;
});
```

### CODE REVIEW

**Comment by islemaster** [comment]
+ @poorvasingal @mrjoshida I believe this covers the `timedLoop()` change you requested this morning?

File: apps/src/lib/util/timeoutList.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/** @file Wrap timer functions for easy cleanup of all timers. */
```
Comment (islemaster) [inline apps/src/lib/util/timeoutList.js:1]: Note: This file was moved, it just also changed enough that it looks like it was added here.

**Comment by poorvasingal** [comment]
Sweet!  Thanks.  @bakerfranke, FYI that we're making this change.  I believe Josh already brought this up with you.

File: apps/src/applab/api.js [hunk 2]
```diff
@@ -388,6 +388,14 @@ export function clearInterval(intervalId) {
                            {'intervalId': intervalId });
 }
 
+export function timedLoop(ms, callback) {
+  return Applab.executeCmd(null, 'timedLoop', {ms, callback});
+}
+
+export function stopTimedLoop(key) {
+  return Applab.executeCmd(null, 'stopTimedLoop', {key});
```
Comment (pcardune) [inline apps/src/applab/api.js:396]: This is not covered by unit tests... https://codecov.io/gh/code-dot-org/code-dot-org/pulls/13360/diff#D14-396

Comment (islemaster) [reply apps/src/applab/api.js:396]: Whoops!  Thanks for catching that.

**Pull request merged**